qemu/tests/virtio_serial_file_transfer.py

```diff
@@ -1,26 +1,222 @@
+    if session is provided
+
+    :param dir_name: where to create the file
+    :param file_size: the size of the file to be created
+    :param session: guest session if have one, perform on host if None
+    :return: the full file path
+    """
+    data_file = os.path.join(dir_name,
+                             "tmp-%s" % utils_misc.generate_random_string(8))
+    cmd = "dd if=/dev/zero of=%s bs=1M count=%d" % (data_file, int(file_size))
+    if not session:
+        error_context.context(
+            "Creating %dMB file on host" % file_size, logging.info)
+        process.run(cmd)
+    else:
+        error_context.context(
+            "Creating %dMB file on guest" % file_size, logging.info)
+        session.cmd(cmd, timeout=600)
+    return data_file
+
+
+@error_context.context_aware
+def _transfer_data(session, host_cmd, guest_cmd, timeout, sender):
+    """
+    Send transfer data command and check result via output
+
+    :param session: guest session
+    :param host_cmd: host script command
+    :param guest_cmd: guest script command
+    :param timeout: timeout for data transfer
+    :param sender: who send data file
+    :return: True if pass, False and error message if check fail
+    """
+    md5_host = "1"
+    md5_guest = "2"
+    try:
+        kwargs = {'cmd': host_cmd,
+                  'shell': True,
+                  'timeout': timeout}
+        error_context.context("Send host command: %s" % host_cmd, logging.info)
+        host_thread = utils_misc.InterruptedThread(process.getoutput,
+                                                   kwargs=kwargs)
+        host_thread.start()
+        g_output = session.cmd_output(guest_cmd, timeout=timeout)
+        if sender == "both":
+            if "Md5MissMatch" in g_output:
+                err = "Data lost during file transfer. Md5 miss match."
+                err += " Script output:\n%s" % g_output
+                return False, err
+        else:
+            md5_re = "md5_sum = (\\w{32})"
+            try:
+                md5_guest = re.findall(md5_re, g_output)[0]
+            except Exception:
+                err = "Fail to get md5, script may fail."
+                err += " Script output:\n%s" % g_output
+                return False, err
+    finally:
+        if host_thread:
+            output = host_thread.join(10)
+            if sender == "both":
+                if "Md5MissMatch" in output:
+                    err = "Data lost during file transfer. Md5 miss "
+                    err += "match. Script output:\n%s" % output
+                    return False, err
+            else:
+                md5_re = "md5_sum = (\\w{32})"
+                try:
+                    md5_host = re.findall(md5_re, output)[0]
+                except Exception:
+                    err = "Fail to get md5, script may fail."
+                    err += " Script output:\n%s" % output
+                    return False, err
```

Comment:
line 111 to line123 and line 126 to line 139 are quite similar, could we make them to a new function?

Reply:
@lijinlijin Updated, please help check, thanks.